libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/kapt/KaptWithoutKotlincTask.kt

```diff
@@ -99,19 +99,24 @@ private class KaptExecution @Inject constructor(
         private const val JAVAC_CONTEXT_CLASS = "com.sun.tools.javac.util.Context"
 
         private fun kaptClass(classLoader: ClassLoader) = Class.forName("org.jetbrains.kotlin.kapt3.base.Kapt", true, classLoader)
+        private var cachedClassLoaderWithToolsJar: ClassLoader? = null
+        private var cachedKaptClassLoader: ClassLoader? = null
     }
 
     override fun run(): Unit = with(optionsForWorker) {
         val kaptClasspathUrls = kaptClasspath.map { it.toURI().toURL() }.toTypedArray()
         val rootClassLoader = findRootClassLoader()
 
-        val classLoaderWithToolsJar = if (toolsJar != null && !javacIsAlreadyHere()) {
+        val classLoaderWithToolsJar = cachedClassLoaderWithToolsJar ?: if (toolsJar != null && !javacIsAlreadyHere()) {
```

Comment:
I have an alternative idea. `tools.jar` should actually be present in a worker's classpath (maybe it's not added now, I don't see if a worker's classpath is modified).
It is not reused probably because of `findRootClassloader` implementation:
```
private fun findRootClassLoader(): ClassLoader {
    tailrec fun parentOrSelf(classLoader: ClassLoader): ClassLoader {
        val parent = classLoader.parent ?: return classLoader
        return parentOrSelf(parent)
    }
    return parentOrSelf(KaptExecution::class.java.classLoader)
}
```
So now the `rootClassLoader` is probably the bootstrap classloader, that does not contain anything besides `rt.jar` classes.
@yanex says that the reasoning behind this was that he was not sure if it's safe to execute KAPT in context of Gradle worker's classloader, because it also contains some Gradle internals, such as Groovy or Guava (see https://github.com/gradle/gradle/issues/3698). However I think it should not be a problem, since KAPT does not use a compiler when running annotation processors.
So I think we should ensure that `tools.jar` and kapt jars are present in a worker's classpath and use a worker's classloader, so up-to-date checks and caching would be effectively handled by Gradle.
If this works, it would be a simpler implementation.
Example implementation: https://github.com/JetBrains/kotlin/commit/97fd98f047701aa4782fc196d814c928fc951075

Reply:
Thanks for reviewing this! I got a similar implementation as your example. Unfortunately, some projects don't build with it. For example, while building CatchUp, antlr throws this:

Caused by: java.lang.UnsupportedOperationException: java.io.InvalidClassException: org.antlr.v4.runtime.atn.ATN; Could not deserialize ATN with UUID 59627784-3be5-417a-b9eb-8131a7286089 (expected aadb8d7e-aeef-4415-ad2b-8204d6cf042e or a legacy UUID).
at org.antlr.v4.runtime.atn.ATNDeserializer.deserialize(ATNDeserializer.java:153)
at androidx.room.parser.SQLiteLexer.(SQLiteLexer.java:777)

I didn't spend much time investigating why it fails and was guessing that the URLClassLoader is a workaround to the problem. Allow me take a deeper look at it.

If the solution to the above turns out to be unfeasible, caching seems to be the way to go. The first implementation I proposed uses HashMap / WeakHashMap. Yan suggested (and I checked with a few projects) that the classloaders created in currently implementation never change (w.r.t. URL and parent). That's why this PR is written. Let me switch back to that implementation, as I also feel it safer, and without measurable performance change compared to this.

Reply:
The assumption that it's safe to execute KAPT in context of Gradle worker's classloader isn't true. There are class name clashes in Gradle Worker's classpath and AP's. In the above case, there are two jars providing antlr, which later finds that the loaded one (from Gradle) is too old.

It's probably better to be resolved in Gradle if that's unwanted behavior (it should, because another worker isolation model CLASSLOADER works just fine). However, that could potentially take a few more weeks and also creates compatibility issues between versions of gradle and kapt. For the time being, caching is probably the way to go?

Reply:
@ting-yuan thanks for investigating! I think current solution is good enough. We can improve later, when/if Gradle fixes worker's classpath. I'll run tests on CI and merge your commit.